src/components/textarea/textarea.tsx

```diff
@@ -47,24 +51,66 @@ export interface TextAreaProps extends RestrictedTextareaHTMLAttributes {
  */
 export const TextArea = forwardRef<HTMLTextAreaElement, TextAreaProps>(
   ({ className, defaultValue, hasError, maxRows = Infinity, minRows = 2, onChange, rows, value, ...rest }, ref) => {
+    const textAreaRef = useRef<HTMLTextAreaElement>(null)
+    const shadowTextAreaRef = useRef<HTMLTextAreaElement>(null)
+
+    // NOTE: We should only use the JS resizing logic if the CSS field-sizing property is not supported
+    const shouldResizeViaJS = !isCSSFieldSizingSupported() && rows === undefined
+
+    useResizeTextAreaEffect({
+      isEnabled: shouldResizeViaJS,
+      shadowTextAreaRef,
+      textAreaRef,
+      value,
+    })
+
+    const decorateOnChange = useResizeTextAreaOnChange({
+      // NOTE: We only want the resizing behaviour to occur on change if the text area should resize AND its
+      // value is uncontrolled.
+      isEnabled: shouldResizeViaJS && value === undefined,
+      shadowTextAreaRef,
+      textAreaRef,
+    })
+
     return (
-      <ElTextArea
-        {...rest}
-        className={cx(hasError && elTextAreaHasError, className)}
-        defaultValue={defaultValue}
-        style={
-          {
-            // NOTE: `rows` takes precedence. If it's defined, the text area is effectively fixed-height.
-            // If it's not defined, we allow resizing to occur between the min and max rows.
-            '--textarea-max-rows': rows ?? maxRows,
-            '--textarea-min-rows': rows ?? minRows,
-          } satisfies TextAreaCSSProperties as CSSProperties
-        }
-        onChange={onChange}
-        ref={ref}
-        rows={rows}
-        value={value}
-      />
+      <>
+        <ElTextArea
+          {...rest}
+          className={cx(hasError && elTextAreaHasError, className)}
+          defaultValue={defaultValue}
+          style={
+            {
+              // NOTE: `rows` takes precedence. If it's defined, the text area is effectively fixed-height.
+              // If it's not defined, we allow resizing to occur between the min and max rows.
+              '--textarea-max-rows': rows ?? maxRows,
+              '--textarea-min-rows': rows ?? minRows,
+            } satisfies TextAreaCSSProperties as CSSProperties
+          }
+          onChange={decorateOnChange(onChange)}
```

Comment:
**note:** here is where we use the new `mergeRefs` function. This allows us to "merge" two refs together to ensure they both receive a handle on the underlying DOM element.